File: 2024/day_15/solution.ts
(() => {
  function drawTable(data: Array<Record<string, string | number>>): string {
    if (!data.length) {
      return "";
    }

    const columns = Object.keys(data[0]);
    const columnWidths = columns.map((col) =>
      Math.max(col.length, ...data.map((row) => String(row[col]).length))
    );

    const border = `+${columnWidths
      .map((width) => "-".repeat(width + 2))
      .join("+")}+`;

    const headerRow = `| ${columns
      .map((col, i) => {
        const cell = col[0].toUpperCase() + col.slice(1);
        return cell.padEnd(columnWidths[i]);
      })
      .join(" | ")} |`;

    const rows = data.map((row) => {
      return `| ${columns
        .map((col, i) => {
          const cell = String(row[col]);
          return cell.padEnd(columnWidths[i]);
        })
        .join(" | ")} |`;
    });

    return [border, headerRow, border, ...rows, border].join("\n");
  }

  console.log(
    drawTable([
      { name: "Alice", city: "London" },
      { name: "Bob", city: "Paris" },
      { name: "Charlie", city: "New York" },
    ])
  );

  console.log(
    drawTable([
      { gift: "Doll", quantity: 10 },
      { gift: "Book", quantity: 5 },
      { gift: "Music CD", quantity: 1 },
    ])
  );
})();

File: 2024/day_15/solution.py
def draw_table(data: list[dict[str, str | int]]) -> str:
    columns = []
    for i, row in enumerate(data):
        if not columns:
            columns = [[c.title(), len(c)] for c in row.keys()]
        j = 0
        for k, v in row.items():
            columns[j][1] = max(len(str(v)), columns[j][1])
            j += 1
    
    divider = "+"
    for col in columns:
        divider += "-" * (col[1] + 2) + "+"
    
    table = []
    table.append(divider)
    
    header = "|"
    for col, width in columns:
        header += " " + str(col).ljust(width) + " |"
    table.append(header)
    
    table.append(divider)
    
    for line in data:
        row = "|"
        c = 0
        for k, v in line.items():
            row += " " + str(v).ljust(columns[c][1]) + " |"
            c += 1
        table.append(row)
    
    table.append(divider)
    
    return '\n'.join(table)
